src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+                {
+                    outEntry.fabric_index = state_info.fabric_index;
+                    outEntry.group        = state_info.group;
+                    outEntry.keyset_index = state_info.keyset_index;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        void Release() override { mProvider.Release(this); }
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric = 0;
+        size_t mPseudoIndex       = 0;
+        size_t mIndex             = 0;
+        size_t mTotalCount        = 0;
+    };
+
+    class KeySetIteratorImpl : public KeySetIterator
+    {
+    public:
+        KeySetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index) :
+            mProvider(provider), mFabric(fabric_index)
+        {
+            FabricStore fabric_store(fabric_index);
+            FabricInfo fabric_info;
+            if (CHIP_NO_ERROR == fabric_store.Load(fabric_info))
+            {
+                mNextId = fabric_info.first_keyset;
+                mCount  = fabric_info.keyset_count;
+                mIndex  = 0;
+            }
+        }
+
+        size_t Count() override { return mCount; }
+
+        bool Next(KeySet & outEntry) override
+        {
+            if (mIndex >= mCount)
+            {
+                return false;
+            }
+
+            KeySetInfo keyset_info;
+            if (CHIP_NO_ERROR != KeySetStore(mFabric, mNextId).Load(keyset_info))
+            {
+                return false;
+            }
+            mIndex++;
+            mNextId                = keyset_info.next;
+            outEntry.keyset_id     = keyset_info.keyset_id;
+            outEntry.policy        = keyset_info.policy;
+            outEntry.num_keys_used = keyset_info.num_keys_used;
+            memcpy(outEntry.epoch_keys, keyset_info.epoch_keys, sizeof(outEntry.epoch_keys));
+            return true;
+        }
+
+        void Release() override { mProvider.Release(this); }
+
+    private:
+        GroupDataProviderImpl & mProvider;
+        chip::FabricIndex mFabric = 0;
+        uint16_t mNextId          = 0;
+        size_t mCount             = 0;
+        size_t mIndex             = 0;
+    };
+
+public:
+    //
+    // Common
+    //
+
+    CHIP_ERROR Init() override
+    {
+#if CHIP_DEVICE_LAYER_TARGET_DARWIN
+        ReturnErrorOnFailure(DeviceLayer::PersistedStorage::KeyValueStoreMgrImpl().Init("chip.store"));
+#elif CHIP_DEVICE_LAYER_TARGET_LINUX
+        ReturnErrorOnFailure(DeviceLayer::PersistedStorage::KeyValueStoreMgrImpl().Init(CHIP_CONFIG_KVS_PATH));
+#endif
+        mInitialized = true;
+        return CHIP_NO_ERROR;
+    }
+
+    void Finish() override { mInitialized = false; }
+
+    //
+    // Group Mappings
+    //
+
+    bool HasGroupNamesSupport() override { return true; }
+
+    bool GroupMappingExists(chip::FabricIndex fabric_index, const GroupMapping & mapping) override
```

Comment:
The logic behind this function seems to elude me. 
According to the spec section 2.5.4 
> A Group Identifier (Group ID or GID) is a 16-bit number that identifies a set of Nodes across a Fabric
at the message layer (see Section 4.12, “Group Key Management”). A Group ID can further be bound
to one or more Endpoints within each Node in the group at the interaction layer.

I would presume that we should retrieve the GroupId from the FabricIdex and then see if we have some endpoints bound to the GroupId and not the other way around.
@bzbarsky-apple any thoughts on this ?

Reply:
Since this is a map, both endpoints to groups, or groups to endpoints works in this case, as long as the (endpoint, group), or (group, endpoint) pair is unique.

Reply:
The direction is:

1 group -> many endpoints

If you do:
- Groups.AddGroup(5) on EP  2, 4, 6, done by Fabric 1
- Groups.AddGroup(5) on EP 2, 4, 10, by Fabric 2

Then Fabric 1's group 5, when a group 5 message comes in from Fabric 1, will dispatch to EP 2,4,6, with accessing fabric 1

Fabric 2's group 5, when a group 5 message comes in from Fabric 2, will dispatch to EP 2,4,10, with accessing fabric 1

Therefore, on EP2, and 4, group ID 5 is used twice, but belongs to two fabrics.